File: app/src/main/java/co/garmax/materialflashlight/features/ForegroundService.java
package co.garmax.materialflashlight.features;

import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.IBinder;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;

import javax.inject.Inject;

import co.garmax.materialflashlight.R;
import dagger.android.AndroidInjection;

/**
 * System won't kill our app with foreground service
 */
public class ForegroundService extends Service {

    private static final int NOTIFICATION_ID = 1;
    private static final String EXTRA_MODE = "extra_mode";

    private static final int MODE_STOP = 0;
    private static final int MODE_START = 1;
    private static final int MODE_LIGHT_OFF = 2;

    @Inject
    LightManager lightManager;

    public static void startService(Context context) {
        context.startService(buildIntent(context, MODE_START));
    }

    public static void stopService(Context context) {
        context.startService(buildIntent(context, MODE_STOP));
    }

    private static Intent buildIntent(Context context, int mode) {
        Intent intent = new Intent(context, ForegroundService.class);
        intent.putExtra(EXTRA_MODE, mode);

        return intent;
    }

    public ForegroundService() {
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        AndroidInjection.inject(this);
        super.onCreate();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        if (intent == null) {
            return super.onStartCommand(intent, flags, startId);
        }

        int mode = intent.getIntExtra(EXTRA_MODE, MODE_STOP);

        // Handle mode
        if (mode == MODE_LIGHT_OFF) {
            lightManager.turnOff();
        } else if (mode == MODE_STOP) {
            stop();
        } else if (mode == MODE_START) {
            startForeground();
        }

        return super.onStartCommand(intent, flags, startId);
    }

    private void stop() {
        stopForeground(true);
        stopSelf();
    }

    // Start foreground service with notification
    private void startForeground() {
        NotificationCompat.Builder notifyBuilder = new NotificationCompat.Builder(getApplicationContext(), "");
        notifyBuilder.setOngoing(true)
                .setSmallIcon(R.drawable.ic_light_notification)
                .setWhen(System.currentTimeMillis())
                .setContentText(getString(R.string.notification_tap_to_turn_off))
                .setContentTitle(getString(R.string.notification_light));

        Intent notificationIntent = buildIntent(getApplicationContext(), MODE_LIGHT_OFF);
        PendingIntent pendingIntent = PendingIntent.getService(getApplicationContext(),
                0,
                notificationIntent,
                PendingIntent.FLAG_CANCEL_CURRENT);
        notifyBuilder.setContentIntent(pendingIntent);

        startForeground(NOTIFICATION_ID, notifyBuilder.build());
    }
}

File: app/src/main/java/co/garmax/materialflashlight/service/ForegroundService.kt
package co.garmax.materialflashlight.service

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.IBinder
import androidx.core.app.NotificationCompat
import androidx.core.app.ServiceCompat
import androidx.core.content.ContextCompat
import co.garmax.materialflashlight.R
import co.garmax.materialflashlight.features.LightManager
import org.koin.android.ext.android.inject
import org.koin.core.context.GlobalContext

/**
 * Keeps the process alive while the light is on.
 */
class ForegroundService : Service() {

    private val lightManager: LightManager by inject()

    override fun onBind(intent: Intent): IBinder? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        intent ?: return super.onStartCommand(intent, flags, startId)

        val command = intent.getIntExtra(EXTRA_COMMAND, COMMAND_STOP)

        if (command == COMMAND_START) {
            startForegroundWithNotification()
            lightManager.turnOn()
            if (!lightManager.isTurnedOn) {
                stop()
            }
        } else {
            lightManager.turnOff()
            stop()
        }

        return START_STICKY
    }

    private fun stop() {
        ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_REMOVE)
        stopSelf()
    }

    private fun startForegroundWithNotification() {
        val stopIntent = Intent(applicationContext, ForegroundService::class.java).apply {
            putExtra(EXTRA_COMMAND, COMMAND_STOP)
        }

        val pendingIntent = PendingIntent.getService(
            this,
            0,
            stopIntent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val notification = NotificationCompat.Builder(this, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_light_notification)
            .setContentTitle(getString(R.string.notification_light))
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setOngoing(true)
            .setSilent(true)
            .setWhen(System.currentTimeMillis())
            .addAction(
                R.drawable.ic_power_off,
                getString(R.string.notification_tap_to_turn_off),
                pendingIntent
            )
            .build()

        createNotificationChannel()

        ServiceCompat.startForeground(
            this,
            NOTIFICATION_ID,
            notification,
            ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE
        )
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                getString(R.string.channel_name),
                NotificationManager.IMPORTANCE_LOW
            ).apply {
                description = getString(R.string.channel_description)
                setShowBadge(false)
            }
            val notificationManager =
                getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.createNotificationChannel(channel)
        }
    }

    companion object {
        private const val NOTIFICATION_ID = 1
        private const val EXTRA_COMMAND = "extra_command"
        private const val CHANNEL_ID = "main"

        private const val COMMAND_STOP = 0
        private const val COMMAND_START = 1

        fun startService(context: Context) {
            val lightManager = GlobalContext.get().get<LightManager>()
            if (!lightManager.ensureRuntimePermissions()) return

            Intent(context, ForegroundService::class.java).apply {
                putExtra(EXTRA_COMMAND, COMMAND_START)
            }.let {
                ContextCompat.startForegroundService(context, it)
            }
        }

        fun stopService(context: Context) {
            Intent(context, ForegroundService::class.java).apply {
                putExtra(EXTRA_COMMAND, COMMAND_STOP)
            }.let {
                context.startService(it)
            }
        }
    }
}
